rework sat policy

guanyilun opened the pull request:
Sorry about the large PR. I have significantly changed the implementation of policy, in a way that's not backwards compatible. It's not fully completed yet, but the PR has become too large that I may prefer to do additional modifications in a new PR when this is merged. It's also a good time to check whether I have the operations implemented correctly. 

Here's a colab notebook I'm using to test this branch:

https://colab.research.google.com/drive/1fs3PR6kEhGtaoFo7dUfMxTGT_hzBOKu7?usp=sharing

Summary of features included in this PR:
- introduce explicit `State` to keep track of operation dependencies
- decouple command definition from policy: introduce operation registration
- proper sun avoidance based on quaternion
- new implementation of `SATPolicy` using a multi-pass approach
- new `SATP1Policy` with defaults
- round-robin calibration target planning

kmharrington commented:
Thanks for that co-lab notebook. it's going to be really helpful totally going through this. Of which I'm just starting but writing up thoughts as I go. 

The first thing I've noticed that you've moved a whole bunch of things into something now called satp1 that should be general to all SATs. Eventually we're going to have at least 6 SATs so we 100% want to have things as general as possible where-ever possible.

We should not have platform specific codes for most of what looks like you've called "operations." (There's a small caveat that _right now_ we need to pass platform specific HWP instance IDs but that should be fixed as soon as the HWP software is in a stable enough configuration for it to be built into sorunlib.) `preamble`, `ufm_relock`, `hwp_spin_up`, `hwp_spin_down`, `det_setup`, `cmb_scan`, `source_scan`, `setup_boresight`, `bias_step`, `wait_until`, `wrap_up`, `set_scan_params` (in which I realize I have just listed every single operation...) none of those should be something that just satp1 knows how to do. Those should be things that all SATs do equally. I think there's a possibility we might want a platform specific `make_operations` function but we might default for now to trying to make it general and only breaking it out if we need it. 

`make_geometry` makes a bit of sense for it to be platform specific because per platform pointing solutions will exist. But we should generalize it somewhat. each SAT has 7 wafers with the same wafer slot numbering (ws0-ws6). The only differences between these should be the shifts in positioning of the wafers. We can also change the config on how that's added. `ws0_shift` would easily be the same name for both platforms (I used `ufm_mv19` in that config I put together because that's how I got the data, but translating that is totally doable), then only the numbers describing the shifts would be different.  

In `make_cal_targets` it would be good to have focus also be just a comma separated list of wafers. we can have those defaults but sometimes we'll probably want just one or two specific wafers. 

I'm also not sure if SATP1State needs to be platform specific? Each instance of an SATState would be specific to a time/platform but I don't understand why we'd need the name of the class to be different.

guanyilun commented:
Thanks for the comments! I started by putting everything as generic as possible but then realized that I don't know what operations for other SATs look like, so I sort of throw everything under SATP1 and wait for people to tell me which one is general to all SATs. It sounds like I should do it the other way around. I'll start moving things

kmharrington commented on src/schedlib/policies/satp1.py:
```diff
@@ -0,0 +1,234 @@
+            },
+        },
+        'operations': operations,
+        'allow_partial': False,
```
it'd be nice to have all of these over-rideable on the creation. I noticed it while wanting to run SATP1Policy.from_defaults but wanting a Jupiter scan. During these time periods Jupiter doesn't actually rise through any of the available "focuses" so it has to be run with allow partial = True. 

Actually... as I think about it. Letting us set allow partial = True for individual cal targets would be the nicest approach.

kmharrington commented on src/schedlib/policies/sat.py:
```diff
@@ -154,62 +251,184 @@
+            "hwp_freq = 2.0",
+            "HWPSpinUp()",
+        ]
+    return state, 0, ["# hwp disabled or already spinning"]
```
The "hwp disabled or already spinning" get's into the comments generated in the schedule file. It's going to make me worried about the HWP state each time. Could you make these return "hwp already spinning" or "hwp disabled" depending on the actual reason for skipping? Same with spinning down.

kmharrington commented:
The other piece I see is that the boresight rotation happens after detector setup. Ex: 
```
################### Detector Setup######################
run.acu.move_to(az=265.0, el=50.0)
run.smurf.take_bgmap(concurrent=True)
run.smurf.iv_curve(concurrent=False, settling_time=0.1)
run.smurf.bias_dets(concurrent=True)
time.sleep(180)
run.smurf.bias_step(concurrent=True)
#################### Detector Setup Over ####################

HWPPrep()
forward = True
hwp_freq = 2.0
HWPSpinUp()
# hwp disabled or already spinning
run.acu.set_boresight(-45.0)   <<<-------------------------------------- Here
run.smurf.bias_step(concurrent=True)
run.seq.scan(
    description='Constant elevation scan.',
    stop_time='2023-10-31T14:04:00+00:00',
    width=40.0, az_drift=0,
    subtype='cmb', tag='265-305',
)
```
If we hadn't already started the telescope at -45 boresight then the detectors would be set up for the wrong positions. 

Also in the co-lab notebook I'm seeing that the boresight is set to 0 degrees for the calibration target but I don't see it change the boresight for that either.

kmharrington commented:
PS: thanks for moving all the operations back to general SATs! Looks a lot better.

guanyilun commented:
Thanks for the very helpful comments! I think I have fixed all of the above points. Most of them are problems with my operation configuration. Scheduler takes a list of operations as argument and adds them in sequential order based on their `sched_mode`. I've now switched the order between `setup_boresight` and `det_setup`. Also, as you suggested, I have added allow_partial as an argument to calibration target. The `setup_boresight` operation not having the right boresight rotation is a slightly deeper problem related to how I plan the operations which took a bit of work to fix. Please feel free to take another look and see if the problems are gone!

guanyilun commented:
I think I'll merging this now. It has become increasingly difficult to maintain this large PR especially when syncing changes from the main branch. This is not backward compatible. For older scheduler please use version <= 0.3.2.
